### PR TITLE
Support Reset factory

### PR DESCRIPTION
**Pull request by Mohammedz93**
#### Why I did it
Support reset factory in Sonic OS
[Reset Factory HLD](https://github.com/sonic-net/SONiC/pull/1231)
[Sonic-mgmt tests](https://github.com/sonic-net/sonic-mgmt/pull/7652)

#### How I did it
- Added new script "/usr/bin/reset-factory"
   * It generates a new config_db.json files with factory configurations
   * It clears system files and logs
   * It removes all docker containers on system except database
   * It clears non-default users and restores default users password
- Dump the default users info to a new file during build "/etc/sonic/default_users.json"
- Supported new type "Keep-basic" in "config-setup factory"
- Add new conf file for config-setup "/etc/config-setup/config-setup.conf

#### How to verify it
- Run reset-factory script with all types: < none | keep-all-config | only-config | keep-basic >
- Run config-setup factory with parameters < none | keep-basic >

#### Description for the changelog
Support reset factory in Sonic OS

### CODE REVIEW

**Comment by liat-grozovik** [comment]
@madhupalu @venkatmahalingam could you please help review the code PR following the HLD review?

**Comment by liat-grozovik** [comment]
@Mohammedz93 please check failures and if relevant to the changes you provided. if not, rerun

**Comment by Mohammedz93** [comment]
@madhupalu @venkatmahalingam could you please review the PR ?

File: files/image_config/config-setup/config-setup [hunk 1]
```diff
@@ -56,17 +57,31 @@ usage()
 EOF
 }
 
+# Factory command usage and help
+usage_factory()
```
Comment (qiluo-msft) [inline files/image_config/config-setup/config-setup:61]: >usage_factory [](http://example.com/codeflow?start=0&length=13)

Could you merge this function to above usage()?

**Comment by qiluo-msft** [comment]
@rajendra-dendukuri Could you help review?

File: files/build_templates/default_users.json.j2 [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{%- set users_dict = {
+                        username: {"password": password,
+                                   "expire": password_expire
+                                  }
```
Comment (qiluo-msft) [inline files/build_templates/default_users.json.j2:4]: >                                   [](http://example.com/codeflow?start=0&length=34)

The indentation in this file is weird. Could you recheck?

File: files/image_config/config-setup/config-setup (hunk 1 above)
Comment (Mohammedz93) [reply files/image_config/config-setup/config-setup:61]: Thanks for the comments.

usage_factory() is used for "config-setup factory" help command.
while useage() is used for "config-setup" help command.
Do you mean by merge, to combine the two functions into one function with parameter ? 
if so, why do you think it is necessary ?

File: files/build_templates/sonic_debian_extension.j2 [hunk 3]
```diff
@@ -586,12 +586,17 @@ sudo bash -c "echo enabled=false > $FILESYSTEM_ROOT/etc/sonic/updategraph.conf"
 # Generate initial SONiC configuration file
 j2 files/build_templates/init_cfg.json.j2 | sudo tee $FILESYSTEM_ROOT/etc/sonic/init_cfg.json
 
-# Copy config-setup script and service file
+# Copy config-setup script, conf file and service file
 j2 files/build_templates/config-setup.service.j2 | sudo tee $FILESYSTEM_ROOT_USR_LIB_SYSTEMD_SYSTEM/config-setup.service
 sudo cp $IMAGE_CONFIGS/config-setup/config-setup $FILESYSTEM_ROOT/usr/bin/config-setup
+sudo mkdir -p $FILESYSTEM_ROOT/etc/config-setup
+sudo cp $IMAGE_CONFIGS/config-setup/config-setup.conf $FILESYSTEM_ROOT/etc/config-setup/config-setup.conf
```
[the remark this answers is not in the document]
Comment (Mohammedz93) [reply files/build_templates/sonic_debian_extension.j2:593]: I’m not sure if I am understanding correctly.
What do you mean by removing the filename ?
"/etc/config-setup" is a folder that was first introduced by by feature: [SONiC-config-setup](https://github.com/sonic-net/SONiC/blob/master/doc/ztp/SONiC-config-setup.md)
It wasn't used until now but I think that this is correct path for locating the conf file
If the concern is that conf file name and folder name have the same prefix, we can change it to  "config_setup.conf"

File: files/build_templates/default_users.json.j2 (hunk 2 above)
Comment (Mohammedz93) [reply files/build_templates/default_users.json.j2]: Fixed, thanks

**Comment by liat-grozovik** [comment]
@qiluo-msft kindly reminder to review following comments handling

**Review by qiluo-msft — APPROVED** [review]

File: files/image_config/reset-factory/reset-factory [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        reset_users
+
+        echo "Delete bash, python and vim history files"
+        find /home /root -type f -name ".bash_history" -o -name ".python_history" -o -name ".viminfo" | xargs rm -rf
+
+        echo "Delete any non-dotfiles in users home directories"
+        find /home/ /root -type f ! -iname ".*" -delete
+    fi
+
+    echo "Remove all docker containers"
```
Comment (qiluo-msft) [inline files/image_config/reset-factory/reset-factory:169]: >Remove all [](http://example.com/codeflow?start=10&length=10)

Do you really remove all?

Comment (Mohammedz93) [reply files/image_config/reset-factory/reset-factory]: Fixed
thanks

File: build_debian.sh [hunk 5]
```diff
@@ -596,6 +596,13 @@ export built_by="$USER@$BUILD_HOSTNAME"
 export sonic_os_version="${SONIC_OS_VERSION}"
 j2 files/build_templates/sonic_version.yml.j2 | sudo tee $FILESYSTEM_ROOT/etc/sonic/sonic_version.yml
 
+# Default users info
+export password_expire="$( [[ "$CHANGE_DEFAULT_PASSWORD" == "y" ]] && echo true || echo false )"
+export username="${USERNAME}"
+export password="$(sudo grep ^${USERNAME} $FILESYSTEM_ROOT/etc/shadow | cut -d: -f2)"
+j2 files/build_templates/default_users.json.j2 | sudo tee $FILESYSTEM_ROOT/etc/sonic/default_users.json
```
Comment (qiluo-msft) [inline build_debian.sh:603]: >default_users.json.j2 [](http://example.com/codeflow?start=25&length=21)

Please check my comment on HLD PR: [https://github.com/sonic-net/SONiC/pull/1231#pullrequestreview-1511603581](https://github.com/sonic-net/SONiC/pull/1231#pullrequestreview-1511603581)

File: build_debian.sh [hunk 6]
```diff
@@ -596,6 +596,13 @@ export built_by="$USER@$BUILD_HOSTNAME"
...
+sudo LANG=c chroot $FILESYSTEM_ROOT chmod 600 /etc/sonic/default_users.json
```
Comment (qiluo-msft) [inline build_debian.sh:604]: >chmod [](http://example.com/codeflow?start=36&length=5)

Also change the user and group of the file to match /etc/shadow

File: build_debian.sh (hunk 5 above)
Comment (Mohammedz93) [reply build_debian.sh:603]: Please see my answer: https://github.com/sonic-net/SONiC/pull/1231#discussion_r1253106720

File: build_debian.sh (hunk 6 above)
Comment (Mohammedz93) [reply build_debian.sh:604]: Done

**Review by qiluo-msft — APPROVED** [review]

**Comment by liat-grozovik** [comment]
@dgsudharsan can you please review as well? if all good, I can move on and merge

**Review by dgsudharsan — APPROVED** [review]

**Comment by dgsudharsan** [comment]
> @dgsudharsan can you please review as well? if all good, I can move on and merge

The changes look good. Please proceed with the merge

**Pull request merged**